File: backend/src/ml/anomalyModel.js
/**
 * Autoencoder inference — dependency-free.
 *
 * The model is a 4-layer MLP of roughly 4k parameters, so running it by hand is
 * ~25 lines of arithmetic. That buys three things worth more than a framework:
 *   - no WASM download in a PWA that has to work offline
 *   - it runs unchanged in Node and in the browser (the on-device claim)
 *   - it ports to a C array for the ESP32 without an intermediate graph format
 *
 * Weights come from ml/export_weights.py. Do not hand-edit model.json.
 *
 * The score is reconstruction error: the autoencoder was trained only on
 * healthy physiology, so a window it cannot reproduce is unlike anything it was
 * shown. High error means "this does not look like normal", NOT "this is
 * disease X" — the model has no diagnostic labels and cannot name a condition.
 */

/** @typedef {{layers: Array, activations: string[], normalisation: {mean: number[], std: number[]}, threshold: number, features: string[], window: number}} ModelSpec */

export function createModel(spec) {
  const { layers, activations, normalisation, threshold, features, window } = spec;
  const { mean, std } = normalisation;

  const nIn = layers[0].in;
  if (nIn !== window * features.length) {
    throw new Error(
      `model.json is inconsistent: first layer takes ${nIn} inputs but window*features = ${window * features.length}`
    );
  }

  // Pre-flatten the weights into typed arrays once, at load.
  const compiled = layers.map((l) => ({
    inDim: l.in,
    outDim: l.out,
    W: Float32Array.from(l.W),
    b: Float32Array.from(l.b),
  }));

  /** y = ReLU?(W x + b), W stored row-major as (out, in). */
  function forward(input) {
    let h = input;
    for (let li = 0; li < compiled.length; li++) {
      const { inDim, outDim, W, b } = compiled[li];
      const out = new Float32Array(outDim);
      for (let o = 0; o < outDim; o++) {
        let sum = b[o];
        const row = o * inDim;
        for (let i = 0; i < inDim; i++) sum += W[row + i] * h[i];
        out[o] = activations[li] === 'relu' ? (sum > 0 ? sum : 0) : sum;
      }
      h = out;
    }
    return h;
  }

  /**
   * @param {Array<Record<string, number|null>>} samples oldest-first, length >= window
   * @returns {{score:number, anomalous:boolean, ratio:number}|null} null when the
   *          window is not yet full or a required vital is missing
   */
  function score(samples) {
    if (!Array.isArray(samples) || samples.length < window) return null;

    const recent = smooth(samples.slice(-window), features);
    if (!recent) return null;
    const x = new Float32Array(window * features.length);

    let k = 0;
    for (let t = 0; t < window; t++) {
      for (let f = 0; f < features.length; f++) {
        const v = recent[t][features[f]];
        if (v === null || v === undefined || Number.isNaN(v)) return null;
        // Same flattening order as training: time-major, features interleaved.
        const s = std[f] || 1e-6;
        x[k] = (v - mean[f]) / s;
        k++;
      }
    }

    const out = forward(x);

    let sum = 0;
    for (let i = 0; i < x.length; i++) {
      const d = out[i] - x[i];
      sum += d * d;
    }
    const err = sum / x.length;

    return {
      score: err,
      anomalous: err >= threshold,
      // How far past the threshold, for display. 1.0 == exactly at it.
      ratio: err / threshold,
    };
  }

  return { score, threshold, features, window, spec };
}

/**
 * Width-5 centred median filter over the window, applied before scoring.
 *
 * The model was trained on BIDMC numerics, where SpO2 is flat 95.7% of the
 * time and HR 72.2% - it effectively learned that these vitals do not move
 * sample-to-sample inside 30 seconds. A real MAX30102 and the simulator both
 * emit integers that flip by one constantly, and the autoencoder cannot
 * reproduce that jitter, so reconstruction error came from sensor noise rather
 * than from physiology. Measured on normal windows (HR 68-84, SpO2 96-98),
 * a one-count SpO2 flicker alone drove 93% of them past the anomaly threshold
 * and 44% past ML_ALERT_RATIO - false alerts on healthy readings.
 *
 * A median is the right filter: it removes single-sample spikes outright while
 * leaving a genuine step edge intact, so a real desaturation still arrives at
 * full amplitude. On the training data itself it is close to a no-op - SpO2 is
 * unchanged in 99.7% of samples - so this conditions serve-time input to look
 * like train-time input rather than introducing a new mismatch. Measured
 * after: 0% false alerts on normal windows, hypoxia still detected 100%.
 *
 * Keep this identical to the smoothing in ml/prepare_data.py.
 */
function smooth(win, features) {
  const n = win.length;
  if (n === 0) return null;
  const half = 2;
  const out = new Array(n);
  const scratch = new Array(2 * half + 1);

  for (let i = 0; i < n; i++) {
    const sample = {};
    for (const f of features) {
      let m = 0;
      for (let j = -half; j <= half; j++) {
        const idx = Math.min(n - 1, Math.max(0, i + j));
        const v = win[idx][f];
        if (v === null || v === undefined || Number.isNaN(v)) return null;
        scratch[m++] = v;
      }
      scratch.sort((a, b) => a - b);
      sample[f] = scratch[half];
    }
    out[i] = sample;
  }
  return out;
}

/**
 * Fixed-length rolling buffer of the vitals the model needs, one per device.
 * Samples with a bad PPG signal are skipped rather than stored — feeding the
 * model values the app itself flags as untrustworthy would manufacture
 * anomalies out of sensor dropout.
 */
export function createBuffer(features, window) {
  const buffers = new Map();

  return {
    push(deviceId, reading) {
      if (reading.signalOk === false) return;

      const sample = {};
      for (const f of features) {
        const v = f === 'HR' ? reading.heartRate : f === 'SpO2' ? reading.spo2 : reading[f];
        if (v === null || v === undefined) return; // incomplete — skip
        sample[f] = v;
      }

      const buf = buffers.get(deviceId) ?? [];
      buf.push(sample);
      if (buf.length > window) buf.splice(0, buf.length - window);
      buffers.set(deviceId, buf);
    },

    get(deviceId) {
      return buffers.get(deviceId) ?? [];
    },

    reset(deviceId) {
      if (deviceId) buffers.delete(deviceId);
      else buffers.clear();
    },
  };
}

File: mobile/lib/core/ml/anomaly_model.dart
import 'dart:convert';
import 'dart:math' as math;
import 'dart:typed_data';

import 'package:flutter/services.dart' show rootBundle;

/// On-device autoencoder inference.
///
/// This is a line-for-line port of `backend/src/ml/anomalyModel.js`. The model
/// is a 4-layer MLP of roughly 4k parameters, so the forward pass is plain
/// arithmetic — no plugin, no native library, no download. It runs entirely on
/// the phone, which is what lets the companion keep scoring vitals with the
/// network off and without a reading ever leaving the device.
///
/// The score is reconstruction error. The autoencoder was trained only on
/// healthy physiology, so a window it cannot reproduce is unlike anything it
/// was shown. A high score means "this does not look like normal" — it is NOT
/// a diagnosis. The model has no labels for named conditions and cannot
/// identify one.
///
/// Weights come from `ml/export_weights.py` via `assets/ml/model.json`. That
/// file is generated; do not hand-edit it.
class AnomalyModel {
  AnomalyModel._({
    required this.features,
    required this.window,
    required this.threshold,
    required List<double> mean,
    required List<double> std,
    required List<_Layer> layers,
  }) : _mean = mean,
       _std = std,
       _layers = layers;

  /// Feature names, in the order the network expects them.
  final List<String> features;

  /// Number of samples the model scores at once.
  final int window;

  /// Reconstruction error at or above which a window is called anomalous.
  final double threshold;

  final List<double> _mean;
  final List<double> _std;
  final List<_Layer> _layers;

  static const String assetPath = 'assets/ml/model.json';

  /// Loads and compiles the model from the bundled asset.
  ///
  /// Returns null when the asset is missing or malformed. The model is
  /// deliberately optional: the rule engine is the floor and must keep working
  /// on its own, so a failure here degrades to "no opinion" rather than
  /// breaking monitoring.
  static Future<AnomalyModel?> load() async {
    try {
      final raw = await rootBundle.loadString(assetPath);
      return fromJson(jsonDecode(raw) as Map<String, dynamic>);
    } on Object {
      return null;
    }
  }

  /// Builds a model from an already-decoded spec. Exposed for tests.
  static AnomalyModel? fromJson(Map<String, dynamic> spec) {
    try {
      final features = (spec['features'] as List).cast<String>();
      final window = (spec['window'] as num).toInt();
      final normalisation = spec['normalisation'] as Map<String, dynamic>;
      final mean = (normalisation['mean'] as List)
          .map((v) => (v as num).toDouble())
          .toList();
      final std = (normalisation['std'] as List)
          .map((v) => (v as num).toDouble())
          .toList();
      final activations = (spec['activations'] as List).cast<String>();

      final layers = <_Layer>[];
      final rawLayers = spec['layers'] as List;
      for (var i = 0; i < rawLayers.length; i++) {
        final l = rawLayers[i] as Map<String, dynamic>;
        layers.add(
          _Layer(
            inDim: (l['in'] as num).toInt(),
            outDim: (l['out'] as num).toInt(),
            weights: Float32List.fromList(
              (l['W'] as List).map((v) => (v as num).toDouble()).toList(),
            ),
            bias: Float32List.fromList(
              (l['b'] as List).map((v) => (v as num).toDouble()).toList(),
            ),
            relu: activations[i] == 'relu',
          ),
        );
      }

      // A mismatch here means the asset and this code disagree about shape,
      // which would otherwise surface as silently meaningless scores.
      if (layers.isEmpty || layers.first.inDim != window * features.length) {
        return null;
      }

      return AnomalyModel._(
        features: features,
        window: window,
        threshold: (spec['threshold'] as num).toDouble(),
        mean: mean,
        std: std,
        layers: layers,
      );
    } on Object {
      return null;
    }
  }

  /// Scores one window, oldest sample first.
  ///
  /// Returns null when the window is short or a required vital is missing —
  /// the model gives no opinion rather than an invented one.
  AnomalyScore? score(List<Map<String, double?>> samples) {
    if (samples.length < window) return null;

    final recent = _smooth(samples.sublist(samples.length - window));
    if (recent == null) return null;

    final n = window * features.length;
    final x = Float32List(n);
    var k = 0;
    for (var t = 0; t < window; t++) {
      for (var f = 0; f < features.length; f++) {
        final v = recent[t][features[f]];
        if (v == null || v.isNaN) return null;
        // Same flattening order as training: time-major, features interleaved.
        final s = _std[f] == 0 ? 1e-6 : _std[f];
        x[k++] = (v - _mean[f]) / s;
      }
    }

    final out = _forward(x);

    var sum = 0.0;
    for (var i = 0; i < n; i++) {
      final d = out[i] - x[i];
      sum += d * d;
    }
    final error = sum / n;

    return AnomalyScore(
      score: error,
      anomalous: error >= threshold,
      ratio: error / threshold,
    );
  }

  /// y = ReLU?(W x + b), with W stored row-major as (out, in).
  Float32List _forward(Float32List input) {
    var h = input;
    for (final layer in _layers) {
      final out = Float32List(layer.outDim);
      for (var o = 0; o < layer.outDim; o++) {
        var sum = layer.bias[o];
        final row = o * layer.inDim;
        for (var i = 0; i < layer.inDim; i++) {
          sum += layer.weights[row + i] * h[i];
        }
        out[o] = layer.relu ? math.max(0.0, sum) : sum;
      }
      h = out;
    }
    return h;
  }

  /// Width-5 centred median filter, applied before normalisation.
  ///
  /// Must stay identical to `smooth()` in anomalyModel.js and `median5()` in
  /// ml/prepare_data.py.
  ///
  /// The model was trained on BIDMC numerics, where SpO2 holds steady between
  /// 95.7% of consecutive samples and HR between 72.2% — it effectively learned
  /// that these vitals do not move inside a 30 second window. A MAX30102 emits
  /// integers that flicker by a count constantly, and a bottlenecked
  /// autoencoder cannot reproduce that jitter, so without this filter the
  /// reconstruction error measures sensor noise instead of physiology. Measured
  /// on windows that were all physiologically normal, a one-count SpO2 flicker
  /// alone pushed 93% of them past the anomaly threshold.
  ///
  /// A median is the right filter here: it removes single-sample spikes
  /// outright while leaving a genuine step edge intact, so a real desaturation
  /// still arrives at full amplitude.
  List<Map<String, double?>>? _smooth(List<Map<String, double?>> win) {
    final n = win.length;
    if (n == 0) return null;
    const half = 2;
    final out = <Map<String, double?>>[];
    final scratch = List<double>.filled(2 * half + 1, 0);

    for (var i = 0; i < n; i++) {
      final sample = <String, double?>{};
      for (final f in features) {
        var m = 0;
        for (var j = -half; j <= half; j++) {
          final idx = math.min(n - 1, math.max(0, i + j));
          final v = win[idx][f];
          if (v == null || v.isNaN) return null;
          scratch[m++] = v;
        }
        scratch.sort();
        sample[f] = scratch[half];
      }
      out.add(sample);
    }
    return out;
  }
}

/// One layer's compiled weights.
class _Layer {
  const _Layer({
    required this.inDim,
    required this.outDim,
    required this.weights,
    required this.bias,
    required this.relu,
  });

  final int inDim;
  final int outDim;
  final Float32List weights;
  final Float32List bias;
  final bool relu;
}

/// The model's opinion on one window.
class AnomalyScore {
  const AnomalyScore({
    required this.score,
    required this.anomalous,
    required this.ratio,
  });

  /// Mean squared reconstruction error.
  final double score;

  /// True when [score] is at or above the model's threshold.
  final bool anomalous;

  /// How far past the threshold, for display. 1.0 is exactly at it.
  final double ratio;
}
